servers/publikator/lib/github/index.js

```diff
@@ -132,22 +145,156 @@ module.exports = {
-        ...commit,
-        repo
-      }))
-      .then(async (commit) => {
-        await redis.setAsync(redisKey, JSON.stringify(commit))
-        return commit
-      })
+
+    if (!rawCommit) {
+      return null
+    }
+    const commit = commitFromGQL(repo, rawCommit)
+    await redis.setAsync(redisKey, JSON.stringify(commit))
+    return commit
+  },
+  getCommits: async (repo, { first = 15, after, before }) => {
+    const { githubApolloFetch } = await createGithubClients()
+    const [login, repoName] = repo.id.split('/')
+    const {
+      data: {
+        repository: {
+          refs: {
+            nodes: heads
+          }
+        }
+      }
+    } = await githubApolloFetch({
+      query: `
+        query repository(
+          $login: String!,
+          $repoName: String!,
+          $maxRefs: Int
+          $maxCommits: Int,
+          $commitsSince: GitTimestamp,
+          $commitsUntil: GitTimestamp
+        ) {
+          repository(
+            owner: $login
+            name: $repoName
+          ) {
+            refs(refPrefix: "refs/heads/", first: $maxRefs) {
+              nodes {
+                name
+                target {
+                  ... on Commit {
+                    oid
+                    author {
+                      date
+                    },
+                    history(first: $maxCommits, until: $commitsUntil, since: $commitsSince) {
+                      pageInfo {
+                        hasNextPage
+                      }
+                      totalCount
+                      nodes {
+                        ... on Commit {
+                          oid
+                          author {
+                            email
+                            name
+                          }
+                          parents (first: 100){
+                            nodes {
+                              oid
+                            }
+                          }
+                          message
+                          committedDate
+                        }
+                      }
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      `,
+      variables: {
+        login,
+        repoName,
+        maxRefs: 100,
+        maxCommits: first,
+        commitsSince: before,
+        commitsUntil: after
+      }
+    })
+    const hasNextPage = heads.some(({target}) => target.history.pageInfo.hasNextPage)
+    const totalCount = heads.reduce((total, {target}) => total + target.history.totalCount, 0)
+
+    const commits = heads
+      .map(({ target }) =>
+        target.history.nodes
+          .map(
+            commit => commitFromGQL(repo, commit)
+          )
+      )
+      .reduce(
+        (acc, v) => acc.concat(v), []
+      )
+      .filter((v, i, arr) => arr.map(mapObj => mapObj.id).indexOf(v.id) === i)
+      .sort((a, b) => descending(a.date, b.date))
+      .slice(0, first)
```

Comment:
If this slices here it potentially needs to change `hasNextPage` to true and set a proper `endCursor` (e.g. if first is 100 and no single branch had more than 100 commits, but 3 had 50)

Reply:
True